### PR TITLE
[ELF] Keep non-alloc orphan sections at the end

### PR DESCRIPTION
**Pull request by MaskRay**
https://reviews.llvm.org/D85867 changed the way we assign file offsets
(alloc sections first, then non-alloc sections).

It also removed a non-alloc special case from `findOrphanPos`.
Looking at the memory-nonalloc-no-warn.test change, which would be
needed by #93761, it makes sense to restore the previous behavior: when
placing non-alloc orphan sections, keep these sections at the end so
that the section index order matches the file offset order.

This change is cosmetic. In sections-nonalloc.s, GNU ld places the
orphan `other3` in the middle and the orphan .symtab/.shstrtab/.strtab
at the end.

### CODE REVIEW

File: lld/ELF/Writer.cpp [hunk 1]
```diff
@@ -1316,8 +1320,8 @@ template <class ELFT> void Writer<ELFT>::sortOrphanSections() {
   i = firstSectionOrDotAssignment;
 
   while (nonScriptI != e) {
-    auto pos = findOrphanPos(i, nonScriptI);
     OutputSection *orphan = &cast<OutputDesc>(*nonScriptI)->osec;
+    auto pos = findOrphanPos(i, nonScriptI);
```
Comment (smithp35) [inline lld/ELF/Writer.cpp:1324]: Although not part of this patch, I'm wondering whether `nonScriptI` would be better called `getNonScriptI` to make it a bit clearer that it is a lambda. It's current name implies that it is a variable.

**Review by smithp35 — APPROVED** [review]
LGTM thanks, may be worth waiting to see if any of the other reviewers have any concerns.

I agree that the difference from GNU ld is cosmetic. As while it may be at the expected section index, its address and file offset are if it was placed at the end.

Output for reference from GNU ld output for `sections-nonalloc.s.tmpb`
```
Section Headers:
  [Nr] Name              Type            Address          Off    Size   ES Flg Lk Inf Al
  [ 0]                   NULL            0000000000000000 000000 000000 00      0   0  0
  [ 1] .text             PROGBITS        00000000000000b0 0000b0 000001 00  AX  0   0  4
  [ 2] .bss              NOBITS          00000000000000b1 0000b1 000001 00  WA  0   0  1
  [ 3] data1             PROGBITS        00000000000000b2 0000b2 000001 00  WA  0   0  1
  [ 4] other1            PROGBITS        00000000000000b4 0000b8 000001 00      0   0  4
  [ 5] other2            PROGBITS        00000000000000b8 0000c0 000001 00      0   0  8
  [ 6] other3            PROGBITS        0000000000000000 0000d0 000001 00      0   0 16
  [ 7] data2             PROGBITS        00000000000000b3 0000b3 000001 00  WA  0   0  1
  [ 8] data3             PROGBITS        00000000000000b4 0000b4 000001 00  WA  0   0  1
  [ 9] .symtab           SYMTAB          0000000000000000 0000d8 000030 18     10   1  8
  [10] .strtab           STRTAB          0000000000000000 000108 000008 00      0   0  1
  [11] .shstrtab         STRTAB          0000000000000000 000110 00004d 00      0   0  1
```

**Review by nga888 — APPROVED** [review]
LGTM.

File: lld/ELF/Writer.cpp (hunk 1 above)
Comment (MaskRay) [reply lld/ELF/Writer.cpp]: Thx. I'll make a separate NFC commit to rename this and reorder the two statements.

**Comment by MaskRay** [comment]
Merged as 9ad0175ea0099351050e1dfb0074d9938f469404. Forgot to use `spr land`

File: lld/ELF/Writer.cpp (hunk 1 above)
Comment (MaskRay) [reply lld/ELF/Writer.cpp]: Hmm, `nonScriptI` is an iterator, not a lambda. So its current name seems fine.
